source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1255,10 +1256,40 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::connPool(
     option->hashKey(hash_key);
   }
 
-  bool have_transport_socket_options = false;
-  if (context && context->upstreamTransportSocketOptions()) {
-    context->upstreamTransportSocketOptions()->hashKey(hash_key);
-    have_transport_socket_options = true;
+  Network::TransportSocketOptionsSharedPtr transport_socket_options =
+      context ? context->upstreamTransportSocketOptions() : nullptr;
+
+  // If configured to do so, we override the ALPN to use for the upstream connection to match the
+  // selected protocol.
+  if (cluster_info_->features() &
+      ClusterInfo::Features::PIN_UPSTREAM_ALPN_TO_SELECTED_HTTP_PROTOCOL) {
+    std::vector<std::string> alpn;
+    switch (protocol) {
+    case Http::Protocol::Http10:
+      alpn.push_back("http/1.0");
+      break;
+    case Http::Protocol::Http11:
+      alpn.push_back("http/1.1");
+      break;
+    case Http::Protocol::Http2:
+      alpn.push_back("h2");
+      break;
+    case Http::Protocol::Http3:
+      alpn.push_back("h3");
+      break;
+    }
+
+    if (transport_socket_options) {
```

Comment:
Doing it here in the cluster manager means that this works for all HTTP connections (including async clients I believe). To do this we wrap the provided transport socket options to override the ALPN that might have been set there - not sure if anything actually sets it here, but this should be more future proof. 

The other idea would be to have this done by the router (router would set it on the transport socket options it provides). This would be a bit cleaner, but would then only apply to calls made through the router